### PR TITLE
Use non-blocking recv() in internal_cancel()

### PR DESCRIPTION
**Pull request by xiong-gang**
The issue of hanging on recv() in internal_cancel() are reported
serveral times, the socket status is shown 'ESTABLISHED' on master,
while the peer process on the segment has already exit. We are not
sure how exactly dose this happen, but we are able to simulate this
hang issue by dropping packet or reboot the system on the segment.

This patch use poll() to do non-blocking recv() in internal_cancel();
the timeout of poll() is set to the max value of authentication_timeout
to make sure the process on segment has already exit before attempting
another retry; and we expect retry on connect() can detect network issue.

Signed-off-by: Ning Yu <nyu@pivotal.io>
Signed-off-by: Gang Xiong <gxiong@pivotal.io>

### CODE REVIEW

**Comment by asimrp** [comment]
Not a concern but a question to clarify my understanding.  The patch is polling before invoking `recv()`.  What if the poll succeeds but a packet sent by the postmaster is dropped.  Can `recv()` still hang?

**Comment by xiong-gang** [comment]
@paul-guo- We considered the way of using `keep-alive`, but we think the interval of `keep-alive` is a little bit hard to control(`tcp_keepalive_intvl`, `tcp_keepalive_probes` and `tcp_keepalive_time`). We need a solid guarantee here: the peer process has exit by the time when the timeout happens. Or else, the peer process on the segment may send `SIGINT` to the QE that has already started on the next query. Similar to this discussion:
https://www.postgresql.org/message-id/flat/20030915070801.GD23844@opencloud.com#20030915070801.GD23844@opencloud.com

**Comment by xiong-gang** [comment]
@asimrp The `recv()` here is to check the exit of the peer process on segments, so i think it's a packet like `FIN`. If the `poll()` succeeded, i think it means the packet already in the receiver's kernel buffer, so the `recv()` should not fail.

File: src/backend/gp_libpq_fe/fe-connect.c [hunk 1]
```diff
@@ -2835,11 +2839,41 @@ internal_cancel(SockAddr *raddr, int be_pid, int be_key,
...
+	pollFds[0].fd = tmpsock;
+	pollFds[0].events = POLLIN;
+	pollFds[0].revents = 0;
+
+	/*
+	 * Wait for at most 660 seconds, which is the sum of max values of
+	 * authentication_timeout and pre_auth_delay. Most likely, it's long enough
+	 * to make sure the process forked by the postmaster on segment is finished.
+	 */
+	pollRet = poll(pollFds, 1, 660 * 1000);
```
Comment (pengzhout) [inline src/backend/gp_libpq_fe/fe-connect.c]: use POLLIN instead of 1?

**Comment by paul-guo-** [comment]
@xiong-gang 

OK. Up to you. Just one thing, for legal retry, i.e. code below,

the timeout should be re-caclcuated (i.e. subtract elapsed time).

+		if (olderrno == EAGAIN || olderrno == EINTR)
  |   |   | +			goto retry5;

**Comment by paul-guo-** [comment]
By the way, do we really need to use max not the current value of (AuthenticationTimeout+ PreAuthDelay)? Hardcoding as 660 seems to be not that good given the code for those values in guc.c could be modified in later version.

**Comment by xiong-gang** [comment]
@paul-guo- 
The reason of using the max value is this GUC could be configured differently on master and segments.
Yes, i should change the hardcoding to const variables.

**Pull request merged**